Ticket1906 check opis using script

Tom-Willemsen opened the pull request:
### Description of work

This is a standalone python script to check OPI formatting. It iterates through OPIs in the C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources directory and checks them against a set of style guidelines.

It is not intended to be a replacement for visually checking an OPI, but rather, as a tool to help find deviations from the template.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/1906

### Acceptance criteria

- [x] The script iterates through the files in the C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources directory
- [x] The script reports how many errors were found in each file
- [x] The script gives no output if a file has no errors
- [x] The script exits with status code 0 if all files are clean, or 1 if any errors are found
- [x] Log files are written in the C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\check_OPI_format_logs directory

The script generates an error under the following conditions:
- [x] Colours of LEDs are not set to the ISIS_* colours
- [x] Plot background colour is not set to the ISIS_* colour
- [x] Text input colours (background/foreground/border) are not the ISIS_* colours
- [x] The fonts used by labels are not the ISIS_* fonts
- [x] LEDs, dropdown menus and push buttons are not within group boxes
- [ ] Labels within grouping containers are not 'Sentence case'
- [ ] Labels outside grouping containers are not 'Title Case'
- [x] Labels within grouping containers do not terminate in a colon (:) or ellipsis (...)
- [ ] Group box headings have every word capitalised except prepositions (Note: the script ignores words of 3 characters or less; these are usually prepositions or abbreviations)

For each of the above conditions:
- [x] There is a (set of) unit test(s)
- [x] The error message is sufficient for a developer to understand what the problem is and how to fix it, but not so verbose that it's annoying.

Known limitations:
- Only checks for LEDs, dropdown menus, push buttons and text inputs outside group boxes - other elements may exist and not throw errors. This should catch the majority of problems in most OPIs, however.
- There will sometimes be valid reasons to violate some of the rules; for example, labels should end in colons, except when used as table headers.
- Doesn't check alignment of objects at all, this will still need to be done by eye.
- Template.opi throws many errors. These are largely due to the yellow annotation labels, which clearly do not conform to the style of an OPI. For testing against a 'known good' OPI, I suggest looking at the eurotherm.

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/GUI-Coding-Conventions)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [x] Are there [system tests](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/System-Testing-with-RCPTT) in place? Do they test a minimal set of functionality and leave the gui as close as possible to its original state?
- [x] Did any existing system test break as a result of the current changes? 
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [x] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has developer documentation been updated if required?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.opis/check_opi_format.py:
```diff
@@ -0,0 +1,272 @@
+            if capitalisation_error:
+                err = "Error on line " + str(name.sourceline) + ": " + etree.tostring(name) \
+                      + "\n... Labels outside grouping containers are titles, and therefore should be 'Title Case'."
+                self.errors.append(err)
```
The previous two methods use the same code. Put it in a separate "check_title_case" method which you call from here with xpath query and error message as arguments. Also allows you to reuse it e.g. if you decide you want to check for title case in additional elements in the future.

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.opis/check_opi_format.py:
```diff
@@ -0,0 +1,272 @@
+                    err = "Error on line " + str(name.sourceline) + ": " + etree.tostring(name) \
+                          + "\n... Labels should be in 'sentence case' " \
+                            "(Expected '" + cased_word + "', but got '" + original_word + "')."
+                    self.errors.append(err)
```
As above, put the code in a "check_sentence_case" method.

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.opis/check_opi_format.py:
```diff
@@ -0,0 +1,272 @@
+                print " --------------\n"
+                build_contains_errors = True
+                for error in self.errors:
+                    print error
```
Write all the results to a txt file rather than (or in addition to) printing it to the console. Also I think it would be nice to have a little summary at the end, like x out of y files had errors.

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.opis/check_opi_format_tests.py:
```diff
@@ -0,0 +1,558 @@
+
+    def test_that_if_a_valid_led_tag_is_parsed_it_causes_no_errors(self):
+        # Arrange
+        checker = CheckOpiFormat()
```
As this line is in every single test, you can instead move it into a setUp-method that automatically gets called before each test.

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.opis/check_opi_format.py:
```diff
@@ -0,0 +1,272 @@
+                filepath = os.path.join(self.root_directory, filename)
+                yield filepath
+
+    def check_plot_area_background(self, root):
```
Why are you checking for plot area background, but not e.g. general foreground / background color?

Tom-Willemsen commented:
I have implemented the fixes you suggested. The logs are now written to the C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\check_OPI_format_logs directory as well as being printed to the console.

ThomasLohnert commented:
Looks good! It's only still missing a check for capitalised first letters in labels before you check for word length, so the checker can catch labels starting with an uncapitalised word >4 characters. 
Finally, it would be great if you could write a little something about the checker in the [OPI creation wiki page](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation) perhaps under the Template OPI section, just mentioning that the checker exists, where it lives and how it should be used etc.

ThomasLohnert commented on base/uk.ac.stfc.isis.ibex.opis/check_opi_format.py:
```diff
@@ -0,0 +1,270 @@
+                if len(word) > self.ignore_short_words_limit \
+                        and not words[0] == words[0].title() \
+                        and not word.lower() == word \
+                        and not any(s in word for s in self.ignore):
```
The functionality is currently broken, as this essentially only checks the first word. As long as that's in title case(`and not words[0] == words[0].title()`), the if-check will always fail (thus, no error)

Tom-Willemsen commented:
Fixed the issues with labels starting with short uncapitalised words and added some notes to the [OPI creation wiki](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation). Also added to the [IOC](https://github.com/ISISComputingGroup/EPICS-ioc/blob/master/PULL_REQUEST_TEMPLATE.md) and [GUI](https://github.com/ISISComputingGroup/ibex_gui/blob/master/PULL_REQUEST_TEMPLATE.md) pull request templates.

mattclarke commented:
Hi Tom. This looks really good. Just a couple of suggestions:

* Could you make it so a different path can be provided at the command line?
* Following on from that, it would be nice to be able to specify a single file to check rather than all of them.

Tom-Willemsen commented:
Added option to scan a single file and specify a full path.

Usage:
`python check_opi_format` Will run as before, iterating through any .opi files in the `C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources` directory

`python check_opi_format Eurotherm.opi` will scan only the Eurotherm opi in the default directory

`python check_opi_format Lakeshore336.opi C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources\Lakeshore336` will scan only the Lakeshore 336 opi in the specified directory.

This information has also been added to the [wiki](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)

DominicOram commented:
Hi Tom,

I'm checking this as part of my CCD100 ticket. Couple of suggestions (sorry):
* Can the default directory be .\resources rather than the full path? My GUI is checked out in a slightly different place so the default doesn't work
* Can you allow the directory for the output file to also be specified?

Otherwise seems to work, certainly brought up lots of errors on my OPI!

Tom-Willemsen commented:
Hi Dom.

I've made the directory a relative path rather than absolute path, as suggested.

I've also made the logs directory location configurable on the command line e.g. `python check_opi_format.py Lakeshore336.opi C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources\Lakeshore336 .\my_logs_folder\` will save the logs in the specified logs folder.

I also added `ISIS` as an ignored word.

DominicOram commented:
Awesome, one more niggle. 
Can you use argparse (https://docs.python.org/3/library/argparse.html) for parsing the command line? It gives you nice things like help and default values. Have a look in block_server.py (https://github.com/ISISComputingGroup/EPICS-inst_servers/blob/master/block_server.py) for an example.

Tom-Willemsen commented:
Now parsing the args using argsparse.

Usage:
`python check_opi_format.py [-h] [-file FILE] [-directory DIRECTORY] [-logs_directory LOGS_DIRECTORY] [-extension EXTENSION]`

e.g. 
`python check_opi_format.py -directory resources/Lakeshore336 -file Lakeshore336.opi -logs_directory my_logs_directory`
will scan a single OPI in `resources/Lakeshore336/Lakeshore336.opi` and save the logs in `/my_logs_directory/`
